ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,243 @@
+        if (getResponse.getCorrelationId() != correlationId) {
+          // The NetworkClient associates a response with a request based on the fact that only one request is sent
+          // out over a connection id, and the response received on a connection id must be for the latest request
+          // sent over it. The check here ensures that is indeed the case. If not, log an error and fail this request.
+          // There is no other way to handle it.
+          logger.error("The correlation id in the GetResponse " + getResponse.getCorrelationId()
+              + " is not the same as the correlation id in the associated GetRequest: " + correlationId);
+          setOperationException(
+              new RouterException("Unexpected internal error", RouterErrorCode.UnexpectedInternalError));
+          operationTracker.onResponse(getRequestInfo.replicaId, false);
+          // we do not notify the ResponseHandler responsible for failure detection as this is an unexpected error.
+        } else {
+          ServerErrorCode getError = getResponse.getError();
+          if (getError == ServerErrorCode.No_Error) {
+            int partitionsInResponse = getResponse.getPartitionResponseInfoList().size();
+            // Each get request issued by the router is for a single blob.
+            if (partitionsInResponse != 1) {
+              setOperationException(new RouterException("Unexpected number of partition responses, expected: 1, " +
+                  "received: " + partitionsInResponse, RouterErrorCode.UnexpectedInternalError));
+              operationTracker.onResponse(getRequestInfo.replicaId, false);
+              // Again, no need to notify the responseHandler.
+            } else {
+              getError = getResponse.getPartitionResponseInfoList().get(0).getErrorCode();
+              responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
+              if (getError == ServerErrorCode.No_Error) {
+                handleBody(getResponse.getInputStream());
+                operationTracker.onResponse(getRequestInfo.replicaId, true);
+              } else {
+                // process and set the most relevant exception.
+                processServerError(getError);
+                if (getError == ServerErrorCode.Blob_Deleted || getError == ServerErrorCode.Blob_Expired) {
+                  // this is a successful response and one that completes the operation regardless of whether the
+                  // success target has been reached or not.
+                  operationCompleted = true;
+                } else {
+                  operationTracker.onResponse(getRequestInfo.replicaId, false);
+                }
+              }
+            }
+          } else {
+            responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
+            operationTracker.onResponse(getRequestInfo.replicaId, false);
+          }
+        }
+      } catch (IOException e) {
```

Comment:
yeah, but catching `Exception` is not the same though, if one is only interested in more specific exceptions.

That said, refactored this because Java 1.7 and later support multi-catch:

`catch (IOException | MessageFormatException e) { ... }`